### ISSUE TITLE
Hoisting

### ISSUE THREAD

**Issue by tomblind** [Perryvw/TypescriptToLua#248]
Got bit by this one:
```ts
foo();
function foo() {}
```

Thanks to hoisting this works just fine in JS but fails in lua. Re-arranging the transpiled code seems like a very non-trivial thing to do but perhaps at least an error from tstl would be good since it can introduce some very confusing bugs.

**Comment by Perryvw** [comment]
Yeah there would have to be some kind of dependency graph resolution to figure that out. Sounds like an interesting problem but for now it is probably easier to just error if the order is problematic.

**Comment by lolleko** [comment]
I don't think we need an dependency graph. Hosting in JS just moves all Declarations to the top of their scope/block irrc.

**Comment by Perryvw** [comment]
What about
```ts
const a = 3;
const b = foo();
function c() { return a; }
```

**Comment by lolleko** [comment]
Hoisting doesnt work for const/let only for var and functions.
Also only the declarations are hoisted:

```
console.log(x);
var x =123;
```
Is valid JS but it will log undefined.

Thats because internally only the declaration is hoisted, teh code above is the same as:
```
var x;
console.log(x);
x = 123;
```

**Comment by Perryvw** [comment]
Aha, that explains it!

**Comment by tomblind** [comment]
It seems there's some edge-cases to this, however.
```ts
function foo() { console.log(x); }
const x = 7;
foo();
```
In JS, this actually shows 7. But in the equivalent lua, it will show nil since x is not declared yet.

**Comment by tomblind** [comment]
So it seems consts, lets and functions exist from the beginning of their block scope, but throw an error if accessed before their declaration. So, properly handling things in lua would involve:
- Declaring all vars at the top of their function scope
- Declaring all lets/consts/functions at the top of their block scope
- Moving all function definitions to after those initial declarations

(horrifying) example:
```ts
function foo(condition: boolean) {
    baz = "baz";
    print(baz);
    if (condition) {
        innerFoo(baz);
        let bar = "bar";
        innerFoo(bar);
        var baz;
        function innerFoo(s: string) {
            print(s);
        }
    }
}
```
```lua
function foo(condition)
    local baz
    baz = "baz";
    print(baz);
    if condition then
        local bar, innerFoo
        innerFoo = function(s)
            print(s);
        end
        innerFoo(baz);
        bar = "bar";
        innerFoo(bar);
    end
end
```

**Comment by tomblind** [comment]
closed by #369
